### ISSUE TITLE
pwclient: allow per-directory settings

### ISSUE THREAD

**Issue by ajdlinux** [getpatchwork/patchwork#38]
Currently, pwclient settings are stored in a single file (~/.pwclientrc), where multiple projects can be defined, but there is only one default.

I use pwclient across a number of projects. pwclient should be able to figure out which project I'm looking at without me always having to explicitly specify it.

I'm thinking something along the lines of:

```
[options]
default=linuxppc-dev

[linuxppc-dev]
url=https://patchwork.ozlabs.org/patchwork/xmlrpc/
dir=/home/ajd/code/linux/

[skiboot]
url=https://patchwork.ozlabs.org/patchwork/xmlrpc/
dir=/home/ajd/code/skiboot*/ # Match a few different skiboot worktrees

[snowpatch]
url=https://patchwork.ozlabs.ibm.com/patchwork/xmlrpc/ # Use our internal test patchwork
dir=/home/ajd/code/snowpatch/
```

Alternatively, a per-directory .pwclientrc, which could perhaps be distributed through the project's git repo.

```
[default]
project=snowpatch
url=https://patchwork.ozlabs.org/patchwork/xmlrpc/
```

(I haven't taken a good look at the pwclient source.)

**Comment by stephenfin** [comment]
Personally, I like the second idea more. Unlike web URLs, the paths to projects often change (at least for me) and do so often. I imagine it would be a nightmare to maintain for some users.

The file used is currently defined [here](https://github.com/getpatchwork/patchwork/blob/0a0f9d549139f235b080915036d53929e9d76294/patchwork/bin/pwclient#L56), while it's used [here](https://github.com/getpatchwork/patchwork/blob/0a0f9d549139f235b080915036d53929e9d76294/patchwork/bin/pwclient#L616). It should be trivial enough to march up from your current directory to root, looking for `pwclientrc` files. If this fails, fallback to the `~` version.

PS: Took me a while to figure out what had changed in your example, so I've updated the description so we have a before/after :wink: 
PPS: You might be interested in [`git-pw`](https://github.com/stephenfin/git-pw). It's not complete (it needs a REST API to be really useful), but it uses git configuration to store project URLs rather than a separate `pwclientrc` file

**Comment by ajdlinux** [comment]
Yeah, on reflection, I think a per-directory setup is ideal.

Thanks for the link to `git-pw` - I'd definitely like to use a patchwork client which integrates more nicely with git than pwclient. Can't wait for the REST API either...

**Comment by kryskool** [comment]
Hi

Why don't use the local configuration from the git repository like

```
git config --local --add patchwork.project snowpatch
git config --local --add patchwork.url https://patchwork.ozlabs.org/patchwork/xmlrpc/
```

to retrieve value (value must be local or global like the existing file ~/.pwclientrc)

`git config --get patchwork.project`

It's per repository, and don't need to add .pwclientrc file in the .gitignore to exclude it

Comments ?

Regards

**Comment by ajdlinux** [comment]
Yeah, that's a good point. It does sound like what Stephen mentioned with the `git-pw` tool.

**Comment by stephenfin** [comment]
We don't do this currently as Patchwork is not SCM specific. As Andrew says though, `git-pw` will do this for us.

**Comment by mirh** [comment]
Ehrm.. It wouldn't hurt if this could still be added. 
It sucks when doing things programmatically.
